[DX-948] Fix order of pages (#3847)
TykTechnologies/tyk-docs: tyk-docs/content/getting-started/key-concepts/what-is-an-api-definition.md

---
date: 2017-03-23T12:53:47Z
Title: API Definition Object
Title: What is Tyk API Definition
description: "Explain the concept of Tyk API definiton"
tags: ["Tyk API definiton", "API definiton", "API definiton Object"]
menu:
  main:
    parent: "Key Concepts"
weight: 55
---

> Tyk handles APIs through files / objects called API Definitions – these are JSON objects – and either live in the `/var/tyk-gateway/apps` directory or in a MongoDB collection as part of a Pro Edition installation.

Reviewer: ```suggestion
Tyk handles APIs through files / objects called API Definitions. These are JSON objects and either live in the `/var/tyk-gateway/apps` directory or in a MongoDB collection as part of a Pro Edition installation.
```

In Tyk, the *Tyk API Definition* is a pivotal configuration object used by the *Tyk API Gateway* to manage and process the inbound requests and outbound responses to and from your actual API service.

> This object is a structured JSON that encapsulates essential details such as the API's name, the listen path (clients' URI for accessing the API via the Tyk Gateway), the target URL (where processed requests are directed, this is the user's API service), and settings for authentication, versioning, and Tyk's rich middleware offerings (out-of-the-box plugins) or custom-user plugins which *Tyk Gateway* can execute or enforce to manage the API traffic. All elements of an API configuration in Tyk are encapsulated in this object and according to it, the *Tyk Gateway* execute them for a certain listen path for request and response, per endpoint (of that listen path), method or globally on the all API endpoints.

Reviewer: ```suggestion
This object is a structured JSON object that encapsulates essential details such as:
- API's name
- Listen path: The clients' URI for accessing the API via the Tyk Gateway.
- Target URL: This is where processed requests are directed to a corresponds to the user's API service.
- Settings for authentication, versioning and Tyk's rich middleware offerings (out-of-the-box plugins) or custom-user plugins which *Tyk Gateway* can execute or enforce to manage the API traffic.

All elements of an API configuration in Tyk are encapsulated in this object and according to it, the *Tyk Gateway* executes them for a certain listen path for request and response. This is done per endpoint (of that listen path), per method or globally across all API endpoints.
```

Author: I thought about that but since these are examples and not a full list I think it's better to keep it as such. In a list people could think that this is the whole list

> This object is a structured JSON that encapsulates essential details such as the API's name, the listen path (clients' URI for accessing the API via the Tyk Gateway), the target URL (where processed requests are directed, this is the user's API service), and settings for authentication, versioning, and Tyk's rich middleware offerings (out-of-the-box plugins) or custom-user plugins which *Tyk Gateway* can execute or enforce to manage the API traffic. All elements of an API configuration in Tyk are encapsulated in this object and according to it, the *Tyk Gateway* execute them for a certain listen path for request and response, per endpoint (of that listen path), method or globally on the all API endpoints.

Author: ```suggestion
This object is a structured JSON object that encapsulates essential details such as the API's name, the listen path (clients' URI for accessing the API via the Tyk Gateway), the target URL (where processed requests are directed, this is the user's API service), the API rate limit and settings for authentication, versioning, and Tyk's rich middleware offerings (out-of-the-box plugins) or custom-user plugins which *Tyk Gateway* can execute or enforce to manage the API traffic. All elements of an API configuration in Tyk are encapsulated in this object and according to it, the *Tyk Gateway* execute them for a certain listen path for request and response, per endpoint (of that listen path), method or globally on the all API endpoints.
```

Reviewer: Maybe could reduce the reduce the length of the sentence, something like...

```suggestion
This object is a structured JSON object that encapsulates essential details such as the API's name, the listen path (clients' URI for accessing the API via the Tyk Gateway), the target URL (where processed requests are directed, this is the user's API service) and the API rate limit. Furthermore, it includes settings for authentication, versioning, and Tyk's rich middleware offerings (out-of-the-box plugins) or custom-user plugins which *Tyk Gateway* can execute or enforce to manage the API traffic.

All elements of an API configuration in Tyk are encapsulated in this object and according to it, the *Tyk Gateway* executes them for a certain listen path for request and response, per endpoint (of that listen path), per method or globally on the all API endpoints.
```

For global configurations of the gateway that are not related to specific user's API, use the [config file]({{<ref "tyk-oss-gateway/configuration">}})

API Definition objects can be quite compact for a basic pass-through API, and can become very complex and large for APIs that require many operations to be completed before a request is proxied.

> ### Two Definition Types

Reviewer: ```suggestion
### API Definition Types
```

> The *Tyk API Definitions* which are written in as JSON objects.

Reviewer: ```suggestion
*Tyk API Definitions* are written as JSON objects.
```

> We have 2 main type of definitions -

Reviewer: ```suggestion
We have two main type of definitions:
```

> We have two main type of definitions:

Reviewer: ```suggestion
We have two main types of definitions:
```

> - **Tyk OAS API definiton** - Our new format, complies with the OpenAPI Spec schema according to the OpenAPI Specification standard. It has a [Tyk OAS API definiton Schema](https://github.com/TykTechnologies/tyk-schemas/blob/main/JSON/draft-04) which you can use in your IDE.

Reviewer: ```suggestion
- **Tyk OAS API definiton**: Our new format, complies with the OpenAPI Spec schema according to the OpenAPI Specification standard. It has a [Tyk OAS API definiton Schema](https://github.com/TykTechnologies/tyk-schemas/blob/main/JSON/draft-04) which you can use in your IDE.
```

> - **Tyk OAS API definiton** - Our new format, comply with the OpenAPI Spec schema according to the OpenAPI Specification standard. It has a [Tyk OAS API definito Schema](https://github.com/TykTechnologies/tyk-schemas/blob/main/JSON/draft-04) which you can use in your IDE.

Reviewer: ```suggestion
- **Tyk OAS API definiton** - Our new format, complies with the OpenAPI Spec schema according to the OpenAPI Specification standard. It has a [Tyk OAS API definiton Schema](https://github.com/TykTechnologies/tyk-schemas/blob/main/JSON/draft-04) which you can use in your IDE.
```

> - **Tyk Classic API definiton** - Tyk original format. It follows a [Tyk Clasic API Definition Schema](https://github.com/TykTechnologies/tyk-schemas/tree/main/JSON/draft-07) which you can use in your IDE.

Reviewer: ```suggestion
- **Tyk Classic API definiton**: Tyk original format. It follows a [Tyk Classic API Definition Schema](https://github.com/TykTechnologies/tyk-schemas/tree/main/JSON/draft-07) which you can use in your IDE.
```

### Object Location

> - For *Tyk OSS*, which includes only *Tyk Gateway* they resied in `/var/tyk-gateway/apps` (LINUX) or `/opt/tyk-gateway/apps` (Docker).

Reviewer: ```suggestion
- For *Tyk OSS*, which includes only *Tyk Gateway* they reside in `/var/tyk-gateway/apps` (LINUX) or `/opt/tyk-gateway/apps` (Docker).
```

> - For *Tyk Self Managed*, the licensed product, they'll be kept in a MongoDB or PostgreSQL.

Reviewer: ```suggestion
- For *Tyk Self Managed*, the licensed product, they'll be kept in a MongoDB or PostgreSQL database.
```

> ### Object Id

Reviewer: ```suggestion
### Object ID
```

API Definitions are identified by their API ID, and Gateway REST calls make reference to this ID where they are used. However, in Dashboard REST calls, an internal ID is used to prevent collisions, and in Dashboard API calls, this API ID must be used when operating on API Configurations.

### Object Fields Documentation
- [Tyk OAS API Definition Objects]({{< ref "tyk-apis/tyk-gateway-api/oas/x-tyk-oas-doc" >}})
- [Tyk Classic API Definition Objects]({{< ref "tyk-gateway-api/api-definition-objects" >}})

### User's API Service and Tyk API Definition

> In contrast to *Tyk API Definition*, the user's actual API service represents the backend service, developed and deployed by users, housing the core business logic and data handling. This service receives requests from the *Tyk API Gateway* after processing based on the *Tyk API Definition*. Crucially, the user's API service remains unaware of the Tyk Gateway's processing layer, responding to incoming requests as in direct client-to-service communication. It implements the API endpoints, resources, and methods providing the service's functionality. It can also have its own OpenAPI document to describe and document itself (which is essencially also another name for *API definition*)

Reviewer: ```suggestion
In contrast to *Tyk API Definition*, the user's actual API service represents the backend service, developed and deployed by users, housing the core business logic and data handling. This service receives requests from the *Tyk API Gateway* after processing based on the *Tyk API Definition*. Crucially, the user's API service remains unaware of the Tyk Gateway's processing layer, responding to incoming requests as in direct client-to-service communication. It implements the API endpoints, resources and methods providing the service's functionality. It can also have its own OpenAPI document to describe and document itself (which is essentially also another name for *API definition*)
```

### Two Definition Types
*Tyk API Definitions* are written as JSON objects.

> We have two main type of definitions:

Author: ```suggestion
We have two main API Definition types:
```

Reviewer: ```suggestion
We have two main API Definition types:
```

Reviewer: ```suggestion
We have two main API Definition types:
```

Reviewer: ```suggestion
We have two main API Definition types:
```

Reviewer: GHub will not allow me to commit, will try resolving and suggesting and commiting in separate comment thread

> We have two main type of definitions:

Reviewer: ```suggestion
We have two main API Definition types:
```

> We have two main type of definitions:

Reviewer: ```suggestion
We have two main API Definition types:
```
